### ISSUE TITLE
Problems with Firefox nightly (57)

### ISSUE THREAD

**Issue by knismaps** [ccloli/E-Hentai-Downloader#78]
Earlier this month with nightly switching over to 57, GreaseMonkey broke since it's not a webextension.  I tried using Tampermonkey and Violentmonkey (which do have webextension versions), but neither seems to allow EHD to work.  I'm not sure if there's an updated version of Greasemonkey to test with yet.  In both, I get a console log like this:

```
16:09:35.526 [EHD] E-Hentai Downloader is running.
16:09:35.526 [EHD] Bugs Report > https://github.com/ccloli/E-Hentai-Downloader/issues | https://greasyfork.org/scripts/10379-e-hentai-downloader/feedback
16:09:35.526 [EHD] To report a bug, it's recommended to provide the logs started with "[EHD]", thanks. =w=
16:09:35.536 [EHD] UserAgent > Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:57.0) Gecko/20100101 Firefox/57.0
16:09:35.536 [EHD] Script Handler > Tampermonkey
16:09:35.536 [EHD] Script Handler Version > 4.4.5484
16:09:35.536 [EHD] E-Hentai Downloader Version > 1.27
16:09:35.536 [EHD] E-Hentai Downloader Setting > {"save-info-list":["title","metas","uploader-comment","page-links"]}
16:09:35.536 [EHD] Current URL > <snip>
16:09:35.536 [EHD] Is Logged In > true
16:09:35.540 [EHD] Request Image Limits From e-hentai.org
16:11:17.891 [EHD] Pages Range > <snip>
16:11:23.691 [EHD] #5: Network Error
16:11:23.691 [EHD] #5: RealIndex > 5  | ReadyState > 4  | Status > 0  | StatusText > 
Request URL > <snip>
ResposeHeaders >
16:11:23.693
[EHD] Index > 5  | RealIndex > 5  | Name > <snip> | RetryCount > 0  | DownloadedCount > 0  | FetchCount > 5  | FailedCount > 0
```
And then for every file it tries, it repeats the last three log lines.  There's no info for status text or response headers.  The log looks the same in both Tampermonkey and Violentmonkey with any gallery I try to download.  The last logline shows a stack as well, but it doesn't seem helpful:
```
failedFetching <snip>:12415:2
onerror <snip>:12900:4
g</< <snip>:2:385
n/< <snip>:30:494
(Async: setTimeout handler) g</< <snip>:2:385
c.safeWindow[b] <snip>:8:40
n <snip>:30:456
J/</< <snip>:36:1
message/< <snip>:19:486
g</< <snip>:2:385
b <snip>:2:232
message <snip>:19:470
message <snip>:20:169
anonymous/< <snip>:70:124
T <snip>:9:268
```
The \<snip\>s here are just the gallery URLs, trying to click to view the JS source doesn't actually work.  failedFetching is in the userscript, I guess.

**Comment by ccloli** [comment]
Works fine on my Firefox Nightly 57.0a1, seems that it's just your network connection is not stable. See #61 , #62 and #63 for more information.

**Comment by knismaps** [comment]
It only stopped working after I had to switch away from GreaseMonkey for 57.  If I use non-nightly with GM, everything is fine.

**Comment by ccloli** [comment]
Looks like it's a bug of GreaseMonkey 4.0alpha1, which doesn't export the GM functions. Tried it on Firefox 55.0.3 but the extension didn't work. 

The nightly version may have some bugs, we recommend you to use the stable channel of Firefox.

**Comment by knismaps** [comment]
I tried with Tampermonkey and Violentmonkey's latest releases, not GM (I didn't easily find a development version of GM that worked with 57 at the time).

**Comment by ccloli** [comment]
Sorry but I didn't get your point. Did you mean one of these?

- The script cannot be run on Nightly 57.0a1, **with GreaseMonkey 4.0alpha1**
    The webextension-based [GreaseMonkey](/greasemonkey/greasemonkey) is still in developing, still not released to public. From my recent test, it didn't behave as what it should be, it still needs some time to migrate.

- The script cannot be run on Nightly 57.0a1, **with Tampermonkey**
    Tested just now with 4.4.5484, it works fine. without throwing any errors or "stopped working". Can you tell me what's the problem did you meet, or show me the output logs again? If the "stopped working" problem is the `Network Error`, then it's just the network error. If it's some others like the script throws an error so that it "stopped working", please copy and paste the error log, and describe how it happens.

**Comment by knismaps** [comment]
Did you mean it _can_ be run with Tampermonkey?

I tried again just now with Tampermonkey 4.4.5484 and the latest nightly and tried to download a gallery.  This resulted in browser logs like I first posted and the status window says "network error."  However, if I just copy and paste the URLs from the logs where it says "request URL >" they all ask me to save/open the images and they download fine.  I tried again to download the gallery with your userscript and it still failed in the same way.  And again, I tried pasting some of the URLs out of the logs and they worked.

**Comment by ccloli** [comment]
Okay I know it. It *really* can be run with Tampermonkey. Sorry to say that it's just network error. 

Maybe you've encountered such a situation, that when you're visiting a gallery online, some images loads very fast, but some are very slow, and sometimes the last pieces won't be load, or even it displays a broken-image icon then. That's the network error, the connection to the server is not stable and it may be broken. 

So why the browser can download it successfully, in fact the browser also has the same situation (not stable connection), but when the connection dropped, the browser will connect to the server again and request the last range of pieces to complete the download. You may not notice it because the browser does it from its side, from your side it just looks like a sequence progress. 

But why the script cannot do it? It should also request the last pieces, instead of saying "Network Error" then starting from the beginning. Yes, the script should be working in that way, but not. That's a technical reason, the script uses `GM_xmlhttpRequest` which implements a cross-origin `XMLHttpRequest`, and from its definition, the script can only get the response data when it finishes the request. That means when the file is transmitting, we don't know what the content is. So we cannot get the unfinished pieces directly, only to wait until it finished, and if it failed, we'll get nothing.

But how the other downloading websites like [Mega](https://mega.nz) that uses XHR to download files and can also pause downloading works? I've explain it in https://github.com/ccloli/E-Hentai-Downloader/issues/5#issuecomment-136361378 , however it may have some grammar mistakes, so let's say that again. 

Mega doesn't works like the script, the script downloads the whole file in only one request, but Mega downloads multiple ranges of file in multiple requests. That's say if the script downloads a 2,000,000 bytes file, it'll just downloads it; Mega will try downloading multiple pieces, like downloading the first 200,000 bytes (0-200,000) then next pieces (200,000-400,000), So Mega can save multiple pieces at the same time, just like IDM (a multiple threads downloading tools). 

If user pauses the download or the connection is dropped, the script has to drop all the downloaded data and fetch again, again, that's a technical reason, we really want to get that pieces to save traffic, but XMLHttpRequest doesn't allow us to do that. How about Mega? The pieces that are still downloading will also be dropped, but the completed pieces will be saved, so they can keep that finished pieces without downloading from the beginning. 

So why not download pieces but a whole file? Well, creating multiple threads to download an image which has a small size (<2MB), the advantage is not that obvious, or even reduce the speed. So this won't be implemented.

So anyway to fix this? As the relative comments say, the most useful way is to use a proxy to improve your speed. Or maybe you should try another program.

**Comment by knismaps** [comment]
I tried with the latest release of Chrome and Tampermonkey (only listed as 4.4), and it works fine.  I don't have the release version of Firefox to compare.  It still doesn't work on nightly.  The settings screen in Tampermonkey for your userscript looks the same in both browsers.  In nightly, it doesn't fail sometimes or go slowly, it just always fails with the same timing.  In Chrome, I only tried to get the first few images of a gallery three times, but it always worked.
